programs/bpf_loader/src/syscalls/cpi.rs

```diff
@@ -879,73 +887,411 @@ fn cpi_common<S: SyscallInvokeSigned>(
+            .unwrap();
+
+        for (new_value, expected_realloc_size) in [
+            (b"foo".to_vec(), MAX_PERMITTED_DATA_INCREASE + 3),
+            (b"foobaz".to_vec(), MAX_PERMITTED_DATA_INCREASE),
+            (b"foobazbad".to_vec(), MAX_PERMITTED_DATA_INCREASE - 3),
+        ] {
+            assert_eq!(caller_account.data, callee_account.get_data());
+            callee_account.set_data_from_slice(&new_value).unwrap();
+
+            update_caller_account(
+                &invoke_context,
+                &memory_mapping,
+                &mut caller_account,
+                &callee_account,
+            )
+            .unwrap();
+
+            let data_len = callee_account.get_data().len();
+            assert_eq!(data_len, *caller_account.ref_to_len_in_vm as usize);
+            assert_eq!(data_len, serialized_len());
+            assert_eq!(data_len, caller_account.data.len());
+            assert_eq!(callee_account.get_data(), &caller_account.data[..data_len]);
+            assert_eq!(data_slice[data_len..].len(), expected_realloc_size);
+            assert!(is_zeroed(&data_slice[data_len..]));
+        }
+
+        callee_account
+            .set_data_length(original_data_len + MAX_PERMITTED_DATA_INCREASE)
+            .unwrap();
+        update_caller_account(
+            &invoke_context,
+            &memory_mapping,
+            &mut caller_account,
+            &callee_account,
+        )
+        .unwrap();
+        let data_len = callee_account.get_data().len();
+        assert_eq!(data_slice[data_len..].len(), 0);
+        assert!(is_zeroed(&data_slice[data_len..]));
+
+        callee_account
+            .set_data_length(original_data_len + MAX_PERMITTED_DATA_INCREASE + 1)
+            .unwrap();
+        assert!(matches!(
+            update_caller_account(
+                &invoke_context,
+                &memory_mapping,
+                &mut caller_account,
+                &callee_account,
+            ),
+            Err(EbpfError::UserError(error)) if error.downcast_ref::<BpfError>().unwrap() == &BpfError::SyscallError(
+                SyscallError::InstructionError(InstructionError::InvalidRealloc)
+            )
+        ));
+    }
+
+    pub type TestTransactionAccount = (Pubkey, AccountSharedData, bool);
+
+    struct MockInvokeContext {
+        transaction_context: TransactionContext,
+        program_accounts: Vec<IndexOfAccount>,
+        preparation: MockInvokeContextPreparation,
+        instruction_data: Vec<u8>,
+    }
+
+    impl MockInvokeContext {
+        fn new(
+            transaction_accounts: Vec<TestTransactionAccount>,
+            instruction_data: impl Into<Vec<u8>>,
+            program_accounts: impl Into<Vec<IndexOfAccount>>,
+            instruction_accounts: &[IndexOfAccount],
+        ) -> Self {
+            let program_accounts = program_accounts.into();
+            let instruction_data = instruction_data.into();
+            let instruction_accounts = instruction_accounts
```

Comment:
This is doing the exact opposite of `prepare_mock_invoke_context()` both cancel out each other. `AccountMeta` is a left over from the old testing framework and should not be used anymore.